### ISSUE TITLE
Core/GameObject: Mature Netherwing Drake

### ISSUE THREAD

**Issue by yalnu** [TrinityCore/TrinityCore#27688]
### Description

**After feeding. The dragon doesn't fly away. And there is no dialog box (Belan shi karkun).**

![WoWScrnShot_013022_014335](https://user-images.githubusercontent.com/93322454/151674532-9d184786-22d7-4731-8893-cf2db5c3b9bb.jpg)

### Expected behaviour

https://www.youtube.com/watch?v=MY1yU4fDNkk

### Steps to reproduce the problem

**Location**
Netherwing Fields in [Shadowmoon Valley](https://tbc.wowhead.com/outland-shadowmoon-valley)
 `X: -4049.340576 Y: 721.523254 Z: 1.619745 Map: 530 (Outland) Area: 3758 (Netherwing Fields) Zone: Shadowmoon Valley`

**Faction Netherwing**
[faction=1015](https://tbc.wowhead.com/faction=1015/netherwing)

**First quest**
[quest=10804](https://tbc.wowhead.com/quest=10804/kindness)

**NPC quest**
[npc=22113](https://tbc.wowhead.com/npc=22113/mordenai)

**NPC drake**
[npc=21648](https://tbc.wowhead.com/npc=21648/mature-netherwing-drake)

**Item quest**
[item=31372](https://tbc.wowhead.com/item=31372/rocknail-flayer-carcass)

**src/server/scripts/Outland/zone_shadowmoon_valley.cpp**

[View git blame](https://github.com/TrinityCore/TrinityCore/blame/50178650b2d1c123c40aa4d7ed3ecf3d7fefa00d/src/server/scripts/Outland/zone_shadowmoon_valley.cpp#L199)

### Branch

3.3.5

### TC rev. hash/commit

4c2986c6b16fac13554d8eb290caa28e863c8361

### Operating system

Ubuntu 20.04.3 LTS

### Custom changes

None

**Comment by anzz1** [comment]
Thank you for your report, it's being fixed.

**Comment by Thor1423** [comment]
@anzz1 @spirehawk 

```c++
Reset();
me->GetMotionMaster()->Clear();
if (uint32 pathId = me->GetWaypointPath())
    me->GetMotionMaster()->MovePath(pathId, true);
```
```sql 
UPDATE `creature_template_addon` SET `AnimTier` = 3 WHERE `entry` = 21648;
UPDATE `creature_template_movement` SET `Flight` = 2 WHERE `CreatureId` = 21648;
```

Simply insert the waypoint, and the dragon will reload and return to its original route.  
and anim 3 makes sure that it doesn't look like the dragon is walking in the air.

https://streamable.com/x3of5p

**Comment by anzz1** [comment]
This is so long ago it's hard to remember, but IIRC the correct way should be to implement the quest in SAI instead of code so it is blizzlike, and IIRC I got it implemented correctly in #27735 but I can't immediately understand why it got closed reading through the notes.
From what I can gather there is issue linked to another issue #23890 with the buffs reseting which hasn't been fixed so the implementation has a hack in it to re-apply the buff after, of course this is not the correct way to do it but #23890 needs to be fixed first as in stop removing (non-periodic-damaging) buffs from non-boss NPCs on evade/combat reset.
So probably closing it was error, as it should have been kept open until #23890 is fixed, and then merged. Also even with the hack to skip the reset bug, its still much better and closer to blizzlike than the current hacky mess in code.
In fact as I mentioned there, I do faintly remember that I did indeed check the implementation against TBC classic that was going on back then, and it was 1:1 blizzlike (except obviously the buff removal->reapply part).

**Comment by Thor1423** [comment]
> This is so long ago it's hard to remember, but IIRC the correct way should be to implement the quest in SAI instead of code so it is blizzlike, and IIRC I got it implemented correctly in [#27735](https://github.com/TrinityCore/TrinityCore/pull/27735) but I can't immediately understand why it got closed reading through the notes. From what I can gather there is issue linked to another issue [#23890](https://github.com/TrinityCore/TrinityCore/issues/23890) with the buffs reseting which hasn't been fixed so the implementation has a hack in it to re-apply the buff after, of course this is not the correct way to do it but [#23890](https://github.com/TrinityCore/TrinityCore/issues/23890) needs to be fixed first as in stop removing (non-periodic-damaging) buffs from non-boss NPCs on evade/combat reset. So probably closing it was error, as it should have been kept open until [#23890](https://github.com/TrinityCore/TrinityCore/issues/23890) is fixed, and then merged. Also even with the hack to skip the reset bug, its still much better and closer to blizzlike than the current hacky mess in code. In fact as I mentioned there, I do faintly remember that I did indeed check the implementation against TBC classic that was going on back then, and it was 1:1 blizzlike (except obviously the buff removal->reapply part).

Whether it's in C++ or SAI doesn't matter.
It's all about the result, and I was able to solve it.

I like SAI too, but unfortunately it's too limited in some aspects, especially when it comes to evading, so that NPCs don't receive their buffs when they're not in combat.

The same problem exists here: https://github.com/TrinityCore/TrinityCore/issues/15420 I built the whole thing in C++ because only C++ allows the buffs to be retained.

It's an SAI problem, and it could take years until the devs fix it, or even consider it important to fix it, so for now, the C++ solution.

**Comment by CraftedRO** [comment]
@Thor1423 the issue is that you tag (@) every past issue participant when most of them are not even longer active or interested in this.

**Comment by anzz1** [comment]
@Thor1423 There was other issue with the C++ script too in this instance it was buggy and not blizzlike in some other ways too. Should have wrote down the details as i don't remember them after 3 years anymore, this is also why i switch to making it properly in SAI, and I compared the end result with TBC classic so it matched.
Yeah bug #23890 needs to be fixed so buffs dont get reset on evade which break lot of scripts and is a major bug. It's not just scripted NPC either but no NPC should lose their (de)buffs when evaded/combat reset, except periodic damage ones (or actually, maybe even periodic damage ones are supposed to stay but since they're out of combat the damage tick just doesn't do anything, need to check against classic if this is the case, with rogue for example this could be done by putting a bleed on a mob and then vanishing and checking if the bleed still stays, iirc it might do but just not do any damage).

I also didn't know originally how things are supposed to be done because it doesn't exactly say it in wiki or anywhere, but the concept of the SAI system is that its a copy of the system blizzard also used to script NPCs, so when you're using it the finer details become more blizzlike unlike in C++ which you can do whatever and basically pull the inner workings out of your ass and unless you confirm the exact working with sniffs that it actually is correct, you very likely end up making un-blizzlike behaviour.

And ultimately like said, my understanding is that blizz make pretty much everything in scripting system like SAI, so if the SAI doesn't work or allow you to do what you need then its bug in SAI itself, or like in this case otherwise in the engine (buffs getting removed).
So this quest would be one of the ones that use some ancient C++ script which is not updated to use SAI yet, which is what needs to be done when updating scripts, removing hacks and fixing it properly.

It's still good that you brought up this issue at is seems to have sat for 3 yrs unnoticed, obviously its sometimes good to up old issues to remember they're still there

**Comment by Thor1423** [comment]
> [@Thor1423](https://github.com/Thor1423) the issue is that you tag (@) every past issue participant when most of them are not even longer active or interested in this.

Because they simply forgot about it, or do you remember every bug report you filed somewhere years ago? I wouldn't remember that.

There are almost 730 bug reports open for 3.3.5, and some are so ancient I hardly think anyone remembers them. Besides, I think they've been forgotten. But some are simple fixes that can be implemented, or even solutions already exist but were never included.

So pushing posts if I have a solution that can be used to make it work is perfectly fine in my opinion.

**Comment by Thor1423** [comment]
> [@Thor1423](https://github.com/Thor1423) There was other issue with the C++ script too in this instance it was buggy and not blizzlike in some other ways too. Should have wrote down the details as i don't remember them after 3 years anymore, this is also why i switch to making it properly in SAI, and I compared the end result with TBC classic so it matched. Yeah bug [#23890](https://github.com/TrinityCore/TrinityCore/issues/23890) needs to be fixed so buffs dont get reset on evade which break lot of scripts and is a major bug. It's not just scripted NPC either but no NPC should lose their (de)buffs when evaded/combat reset, except periodic damage ones (or actually, maybe even periodic damage ones are supposed to stay but since they're out of combat the damage tick just doesn't do anything, need to check against classic if this is the case, with rogue for example this could be done by putting a bleed on a mob and then vanishing and checking if the bleed still stays, iirc it might do but just not do any damage).
> 
> I also didn't know originally how things are supposed to be done because it doesn't exactly say it in wiki or anywhere, but the concept of the SAI system is that its a copy of the system blizzard also used to script NPCs, so when you're using it the finer details become more blizzlike unlike in C++ which you can do whatever and basically pull the inner workings out of your ass and unless you confirm the exact working with sniffs that it actually is correct, you very likely end up making un-blizzlike behaviour.
> 
> And ultimately like said, my understanding is that blizz make pretty much everything in scripting system like SAI, so if the SAI doesn't work or allow you to do what you need then its bug in SAI itself, or like in this case otherwise in the engine (buffs getting removed). So this quest would be one of the ones that use some ancient C++ script which is not updated to use SAI yet, which is what needs to be done when updating scripts, removing hacks and fixing it properly.
> 
> It's still good that you brought up this issue at is seems to have sat for 3 yrs unnoticed, obviously its sometimes good to up old issues to remember they're still there

The problem is that features missing from SAI are considered low priority. I also have a few wishes to make certain functions compatible with SAI, but that won't happen because, according to the statement, SAI is only intended for trash NPCs, not for completing complex tasks.

And before I wait for https://github.com/TrinityCore/TrinityCore/issues/23890 to be fixed, I'll first look for other solutions, and you can find them in pure C++ programming.

But using C++ instead of SAI isn't a direct hack per se, but just a different method. I like SAi; I designed my custom dungeons and raids entirely in SAI. But I had to take a lot of detours because SAI lacks too many features.

And I'm not going to wait decades for the quest to work. If SAI ever gets the necessary features, which I doubt so far based on what some devs have said, then I can always rebuild it, but until then I'll suggest alternative solutions.

**Comment by anzz1** [comment]
Yeah to my understanding this is what the theory is on how Blizz have done it, that SAI-type system is used for most of the scripting, quests. dungeons, enemies, and such, while the most complex interactions on raid bosses and such would be one-off hand-made things in C++, and this is also the system that TrinityCore aims to replicate. This includes replicating the limitations of SAI, as using similarly-limited scripting system as the original is how to create blizzlike-behaviour. Of course there is still a lot of guesswork involved on how it might have been implemented as sniffs can only confirm the end-result matching (which is ultimately the thing that matters), not the underlying server-side system. Still, from the variables involved the TC team have been able to make good educated guesses on this, and implement the system to be as close as what it originally would be.

And in this particular case, if you look at my PR I did already implement this quest in SAI so its not necessary to use C++ in this case. I think the merge request on that was probably mistakenly closed, maybe @offl could weigh in on that if he remembers as he seems to have been active recently in this project.

Note though that this is my understanding on the matter when I went through some of the history of the project to teach myself on the project, I might not be entirely correct on some of these claims and obviously am just a small-time contributor and not a maintainer so my word shouldn't be taken as gospel.

**Comment by Thor1423** [comment]
I didn't mean any harm, I just wanted to help make the whole thing more Blizzard-like. Since it's currently still implemented in C++, I experimented a bit and came up with this result.

With this change, the dragons fly back as intended and don't stay on the ground like shown in the screenshot. This was just a suggestion on my part.

**Comment by anzz1** [comment]
Yes of course, didn't mean to imply that at all. It's good that you lifted this issue especially because it links to another important issue of mobs having their buffs reset on evade which should not happen. I hope the maintainers could weigh in on this, how to proceed.

**Comment by Thor1423** [comment]
> Yes of course, didn't mean to imply that at all. It's good that you lifted this issue especially because it links to another important issue of mobs having their buffs reset on evade which should not happen. I hope the maintainers could weigh in on this, how to proceed.

I don't push old posts if I can't offer a solution, whether it's with SAI or C++, it doesn't matter. I don't want to know how many threads here in the bug tracker have already been forgotten because people no longer believe they'll be fixed.

Some bugs are so trivial that they're easy to fix, and some require a bit more work and testing, but if you post a possible solution, we can discuss it and talk about it together.

In my case, I preferred C++ because I know that SAI doesn't currently have this option, and until it's introduced at some point, you can update the C++ code so that it at least works properly.

Therefore, implementing my solution would be a better choice than waiting X years until SAI is ready. There are quests that don't work at all and thus block an entire quest chain or possibly even a good story quest series.

I just finished fixing a quest that's now almost Blizzlike, but a few details are still missing. I had to replace a simple SAI script with a C++ script, and why? Because another function doesn't work in SAI, namely excluding certain NPCs so you don't get aggro from them, but all other functions remain as they are. I also tried it with MoveLineOfSight, but no matter what I chose, it didn't work.

Example code:

```c++
void MoveInLineOfSight(Unit* who) override
{
if (!who)
return;

if (who->GetEntry() == 12345)
return;
}
```

This tells me that NPC 12345, when I walk past him in my vehicle NPC, I don't get aggro, and that's Blizzlike in the quest area.

So I have two options: either I leave it as it is and wait until SAI can do it, or I give it a simple C++ script so it works the way it's supposed to.

My custom folder is already full of scripts where I've fixed quests/NPCs because SAI can't do it that way or it simply worked better with C++.

I use chatgpt + TC source and specify what I need and what the AI ​​can access to create functions. Sure, the code might not be as perfect as it would be from a real programmer, but it works and delivers the result I need.
